roomserver/state/state.go

```diff
@@ -681,6 +681,60 @@ func (v StateResolution) calculateStateAfterManyEvents(
 	return
 }
 
+// ResolveConflictsAdhoc is a helper function to assist the query API in
+// performing state resolution when requested. This is a different code
+// path to the rest of state.go because this assumes you already have
+// gomatrixserverlib.Event objects and not just a bunch of NIDs like
+// elsewhere in the state resolution.
+// TODO: Some of this can possibly be deduplicated
+func ResolveConflictsAdhoc(
+	version gomatrixserverlib.RoomVersion,
+	events []gomatrixserverlib.Event,
+	authEvents []gomatrixserverlib.Event,
+) ([]gomatrixserverlib.Event, error) {
+	type stateKeyTuple struct {
+		Type     string
+		StateKey *string
+	}
+
+	eventMap := make(map[stateKeyTuple][]gomatrixserverlib.Event)
```

Comment:
Fairly sure this doesn't work because `*string` won't be dereferenced when checking key equality. See https://play.golang.org/p/peYVWfZLtA7

Reply:
You're right, that's a mistake.

Reply:
Fixed now and added comments.